### PR TITLE
Add -fdiagnostics-show-caret to display the context of an error

### PR DESCRIPTION
**Pull request by lefessan**
When a warning or error is printed, prints the source context with the line of the error, plus 2 lines before and after.
COBC_PREVIEW_ERROR can be used to set the option globally.

### CODE REVIEW

File: tests/testsuite.src/used_binaries.at [hunk 1]
```diff
@@ -793,19 +793,40 @@ AT_CLEANUP
...
        PROGRAM-ID.      prog.
        PROCEDURE        DIVISION.
            DISPLAY "jobby"
            EXIT HERE.
 ])
+AT_CHECK([$COMPILE -j prog.cob], [1], [], [prog.cob:6: error: syntax error, unexpected Identifier
+])
+
+# check -fpreview-error prints error context
+AT_CHECK([$COMPILE -j -fpreview-error prog.cob], [1], [], [prog.cob:6: error: syntax error, unexpected Identifier
```
Comment (lefessan) [inline tests/testsuite.src/used_binaries.at:808]: Here is an example...

**Comment by lefessan** [comment]
As an example:
```
prompt> $COMPILE -j -fpreview-error prog.cob
prog.cob:6: error: syntax error, unexpected Identifier

  0004          PROCEDURE        DIVISION.
  0005              DISPLAY "jobby"
  0006 >            EXIT HERE.

```

**Comment by lefessan** [comment]
Note that the default is currently to disable this feature, but it could be activated by default (and disabled through `COBC_PREVIEW_ERROR=0` for the testsuites)

**Comment by GitMensch** [comment]
Hm, the idea is good and the general point of reworking the diagnostic format is even on the GSOC idea list for GnuCOBOL (did you get it from there/the note on the dev list about that?).

But reading the source file again should be absolute last resort and will be very likely wrong for REPLACE/REPLACING and won't work for complies from stdin.

So... Could you please rework that to possibly save the context from the scanner via to a static buffer?

Also note that -fno-diagnostics-show-line-numbers should be added along with this change (which would drop the leading line numbers) and possibly -fdiagnostics-show-caret should be used to toggle this instead of -fpreview-error ... and ideally we should also use the caret to show where the error is.

As a general rule of thumb we'd like to reach https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Message-Formatting-Options.html ...
someday.

**Comment by lefessan** [comment]
> Hm, the idea is good and the general point of reworking the diagnostic format is even on the GSOC idea list for GnuCOBOL (did you get it from there/the note on the dev list about that?). 

No, I didn't know about the GSOC idea. Is it discussed on the forum ?
I did that patch because I needed it myself... which is usually how I work, as I can only devote a small amount of time on the project from time to time.

> But reading the source file again should be absolute last resort and will be very likely wrong for REPLACE/REPLACING and won't work for complies from stdin. So... Could you please rework that to possibly save the context from the scanner via to a static buffer? 
> Also note that -fno-diagnostics-show-line-numbers should be added along with this change (which would drop the leading line numbers) and possibly -fdiagnostics-show-caret should be used to toggle this instead of -fpreview-error ... and ideally we should also use the caret to show where the error is. 

I won't be able to spend much more time on this patch. For now, I changed the name of the option to `-fdiagnostics-show-caret` and added another option to disable it for warnings (typically, we are working on a project where there are currently a lot of warnings, and we want to only focus on errors in a first step).

> As a general rule of thumb we'd like to reach https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Message-Formatting-Options.html ... someday.

Would it be ok to merge this work as it is, and wait for the GSOC to get it improved ? I will be working on adding a bit more GCOS support now, so I won't be able to improve this patch more before some time.

**Comment by lefessan** [comment]
@GitMensch Hi, I say on the forum that you mentioned this patch. Do you think it would be ok to merge it in the current state ?

File: cobc/error.c [hunk 2]
```diff
@@ -63,10 +70,73 @@ print_error_prefix (const char *file, int line, const char *prefix)
...
+   before any REPLACE.
+ */
+static void
+diagnostics_show_caret  (enum cb_error_kind kind, const char *file, int line)
+{
+	if ( cb_diagnostics_show_caret
+	     && file != NULL && line
+	     && ( cb_diagnostics_warning_show_caret || kind == CB_KIND_ERROR )
+		){
+		FILE* fd = fopen(file, "r");
```
Comment (GitMensch) [inline cobc/error.c:88]: needs to check `fd` after open and likely should check before and not do that if name matches `COB_DASH`

**Review by GitMensch — CHANGES_REQUESTED** [review]
> @GitMensch Hi, I've seen on the forum that you mentioned this patch. Do you think it would be ok to merge it in the current state ?

Not in its current state :-/ But after reviewing the [initial GCC implementation](https://github.com/gcc-mirror/gcc/commit/9fec00429dd36ea51798afeaf5f6b9c1987b85cd) (also opened the file then read and print the lines), I'm now open to add it without the design change I previously thought of (GCC did that later on).

Things that are likely easy to do:

* drop `COBC_DIAGNOSTICS_SHOW_CARET` completely; for the testsuite it is enough to add `-fno-diagnostics-show-caret` in tests/atlocal.in, just as I did with `-fno-diagnostics-show-option`
* rebase
* add `-fno-diagnostics-show-line-numbers` (because it is just a single flag that only modifies the output in one place)
* add an entry in NEWS and doc/gnucobol.tex about the two new options

Then, as a follow up, we may [use source caches, like the current GCC implementation does](https://github.com/gcc-mirror/gcc/blob/5d85b5d649fff675ff00adcc99371bccf4ef5944/gcc/input.cc#L858).

But still you may want to have a look at the old diagnostic print function linked before your commit and look for good style; and have a look at [the current one](https://github.com/gcc-mirror/gcc/blob/5d85b5d649fff675ff00adcc99371bccf4ef5944/gcc/diagnostic-show-locus.cc#L2831)

File: cobc/cobc.c [hunk 3]
```diff
@@ -8648,6 +8648,18 @@ set_cobc_defaults (void)
 		 *p == '1')) {
 		cb_unix_lf = 1;
 	}
+	p = cobc_getenv ("COBC_DIAGNOSTICS_SHOW_CARET");
```
[the remark this answers is not in the document]
Comment (lefessan) [reply cobc/cobc.c:8662]: I want to argue to keep it: most of the options in the compiler do not need an env variable because they are project-dependent, i.e. the project developers should whether these options should be set or not for their project. On the contrary, the way errors are displayed is user-dependent, not project-dependent. As such, users may want to set or unset these options in their config files, so that they are used by `make` automatically. So the env variable is needed for that (unless we add a user config file for gnucobol).

File: cobc/flag.def [hunk 4]
```diff
@@ -232,6 +232,16 @@ CB_FLAG (cb_listing_symbols, 1, "tsymbols",
...
+CB_FLAG_ON (cb_diagnostics_show_option, 1, "diagnostics-show-option",
 	_("  -fno-diagnostics-show-option\tsuppress output of option that directly\n"
 	  "                        controls the diagnostic"))
+
+CB_FLAG (cb_diagnostics_show_caret, 1, "diagnostics-show-caret",
+	_("  -fdiagnostics-show-caret     display source context on warning/error diagnostic\n"
+	  "                               (the source is displayed without any REPLACE performed."
+	))
+
+CB_FLAG_ON (cb_diagnostics_warning_show_caret, 1, "diagnostics-show-caret-on-warning",
```
[the remark this answers is not in the document]
Comment (lefessan) [reply cobc/flag.def]: I removed it.

File: cobc/error.c (hunk 2 above)
Comment (lefessan) [reply cobc/error.c]: I added checks for NULL and COB_DASH.

File: cobc/cobc.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (GitMensch) [reply cobc/cobc.c:8662]: I see the point - but we shouldn't do this for a single option, a user may want to have the diagnostic name or not, he maybe even don't care about any warnings, ...
A project specific setup that handles this would set default flags in the makefile and passes additional flags from the user's environment or include file (depending on the setup either before or after the default makefile flags).

Actually the way projects can nicely be setup via Makefiles is the main reason for not having a global env var or a config file (other than dialect options). There's a FR about this on SF if you want to give that more thoughts - for this PR, please drop that.

File: NEWS [hunk 5]
```diff
@@ -325,12 +325,19 @@ NEWS - user visible changes				-*- outline -*-
...
+  the error message.  With -fno-diagnostics-show-line-numbers, the line
+  numbers are not printed within the content.
+
 * Important Bugfixes:
 
 ** for dialects other than the GnuCOBOL default different reserved "alias" words
-   were not usable, for example SYNCHRONIZED or COMPUTATIONAL. This was fixed
-   and reserved words updated for the dialects "acu" (to ACUCOBOL-GT 10.4),
-   "ibm" (to Enterprise COBOL 6.3) and "mf" (to Micro Focus Visual COBOL 6.0)
+   were not usable, for example SYNCHRONIZED or COMPUTATIONAL. This
```
Comment (GitMensch) [inline NEWS:337]: unrelated change?

File: cobc/error.c [hunk 6]
```diff
@@ -63,10 +70,80 @@ print_error_prefix (const char *file, int line, const char *prefix)
...
+		}
+		buffer[char_pos] = c ;
+		if (c == '\n' || char_pos == CARET_MAX_COLS){
+			buffer[char_pos] = 0;
+			if (line_pos > line-3 && line_pos < line+3){
+				if (line_pos == line-2) fprintf(stderr, "\n");
+				printed = 1;
+				fprintf (stderr, "  ");
+				if (cb_diagnostics_show_line_numbers){
+					fprintf (stderr, "%04d ",
```
Comment (GitMensch) [inline cobc/error.c:108]: Currently that does not only skip line number but the whole printing, no?

File: tests/testsuite.src/used_binaries.at [hunk 7]
```diff
@@ -918,3 +912,100 @@ AT_CHECK([cat prog.cob | $COMPILE_MODULE -j -], [0], [job], [],
...
+
+  0005          WORKING-STORAGE  SECTION.
+  0006          01 TEST-VAR PIC 9(2) VALUE 'A'.
+  0007 >        COPY 'CRUD.CPY'.
+  0008          PROCEDURE        DIVISION.
+  0009              DISPLAY TEST-VAR NO ADVANCING
+
+prog.cob:6: warning: numeric value is expected
+])
+AT_CHECK([$COMPILE -fdiagnostics-show-caret -fno-diagnostics-show-caret-on-warning -j prog.cob], [1], [], [prog.cob:7: error: CRUD.CPY: No such file or directory
```
Comment (GitMensch) [inline tests/testsuite.src/used_binaries.at:999]: `-fno-diagnostics-show-caret-on-warning` doesn't exists any more, please replace it with a test for `-fno-diagnostics-show-line-numbers`

**Review by GitMensch — CHANGES_REQUESTED** [review]
Nearly ready now - see review details.

**Comment by GitMensch** [comment]
freindly ping @lefessan

File: tests/testsuite.src/used_binaries.at (hunk 7 above)
Comment (lefessan) [reply tests/testsuite.src/used_binaries.at]: done, replaced with a test on `-fno-diagnositcs-show-line-numbers`

File: cobc/error.c (hunk 6 above)
Comment (lefessan) [reply cobc/error.c:108]: Wrong indentation, sorry...

**Comment by lefessan** [comment]
I think all your requests have been fixed.

File: cobc/flag.def [hunk 8]
```diff
@@ -232,6 +232,16 @@ CB_FLAG (cb_listing_symbols, 1, "tsymbols",
...
 	_("  -fno-diagnostics-show-option\tsuppress output of option that directly\n"
 	  "                        controls the diagnostic"))
+
+CB_FLAG (cb_diagnostics_show_caret, 1, "diagnostics-show-caret",
+	_("  -fdiagnostics-show-caret     display source context on warning/error diagnostic\n"
+	  "                               (the source is displayed without any REPLACE performed."
+	))
+
+CB_FLAG_ON (cb_diagnostics_show_line_numbers, 1, "diagnostics-show-line-numbers",
+	_("  -fno-diagnostics-show-line-numbers     suppress display of line numbers in diagnostics"
```
Comment (GitMensch) [inline cobc/flag.def:245]: "strange" thing to do, but please start the second line on the same position that is used for `diagnostics-show`... and as the name is long delimit the first line with \t instead of spaces

File: tests/testsuite.src/used_binaries.at [hunk 9]
```diff
@@ -794,23 +794,17 @@ AT_CLEANUP
 
 AT_SETUP([run job with unsuccessful compilation])
 AT_KEYWORDS([runmisc cobcrun])
-
-
-AT_DATA([prog.cob], [
+AT_DATA([prog.cob],[
```
Comment (GitMensch) [inline tests/testsuite.src/used_binaries.at:797]: please drop these and the following unrelated changes (enough during commit)

File: tests/testsuite.src/used_binaries.at [hunk 10]
```diff
@@ -918,3 +912,62 @@ AT_CHECK([cat prog.cob | $COMPILE_MODULE -j -], [0], [job], [],
...
+       COPY 'CRUD.CPY'.
+       PROCEDURE        DIVISION.
+           DISPLAY TEST-VAR NO ADVANCING
+           END-DISPLAY
+           MOVE 12 TO TEST-VAR
+           DISPLAY TEST-VAR NO ADVANCING
+           END-DISPLAY
+           STOP RUN.
+])
+AT_CHECK([$COBC -Wall -fsyntax-only prog.cob], [1], [], [prog.cob:7: error: CRUD.CPY: No such file or directory
```
Comment (GitMensch) [inline tests/testsuite.src/used_binaries.at:935]: this line should be:

```
AT_CHECK([$COMPILE_ONLY prog.cob], [1], [],
[prog.cob:7: error: CRUD.CPY: No such file or directory
```

the following test should be

```
AT_CHECK([$COMPILE_ONLY -fdiagnostics-show-caret -fdiagnostics-show-line-numbers prog.cob], [1], [],
[prog.cob:7: error: CRUD.CPY: No such file or directory
```

and the last (which may be switched with the second)

```
AT_CHECK([$COMPILE_ONLY -fdiagnostics-show-caret -prog.cob], [1], [],
[prog.cob:7: error: CRUD.CPY: No such file or directory
```

File: cobc/flag.def [hunk 11]
```diff
@@ -232,6 +232,16 @@ CB_FLAG (cb_listing_symbols, 1, "tsymbols",
 CB_FLAG (cb_listing_cmd, 1, "tcmd",
 	_("  -ftcmd                specify command line in listing"))
 
-CB_FLAG_ON (cb_diagnostic_show_option, 1, "diagnostics-show-option",
+CB_FLAG_ON (cb_diagnostics_show_option, 1, "diagnostics-show-option",
 	_("  -fno-diagnostics-show-option\tsuppress output of option that directly\n"
 	  "                        controls the diagnostic"))
+
+CB_FLAG (cb_diagnostics_show_caret, 1, "diagnostics-show-caret",
```
Comment (GitMensch) [inline cobc/flag.def:239]: If possible please change to `CB_FLAG_ON` and add the reverse to the flags in tests/atlocal.in

**Review by GitMensch — CHANGES_REQUESTED** [review]
only minor changes needed (feel free to directly commit)

If you want to finish that completely, then add `-fdiagnostics-plain-output` to cobc.c (this is a "normal" option, no flag, it has no "no" and would set both new options to ´0`) and use this in tests/atlocal.in for the flags.

**Comment by GitMensch** [comment]
Possibly use in the NEWS something similar to GCCs entry:

the diagnostics now print the source code context with a left margin showing line numbers, configurable with -fno-diagnostics-show-line-numbers, and possible to disable completely with -fno-diagnostics-show-caret;

the option -fdiagnostics-plain-output was added to request that diagnostic output look as plain as possible and stay more stable over time

File: cobc/error.c [hunk 12]
```diff
@@ -63,10 +70,80 @@ print_error_prefix (const char *file, int line, const char *prefix)
 	}
 }
 
+/* Display a context around the location of the error/warning, only if
+ * cb_diagnostics_show_caret is true
+
+   Only display two lines before and after. No caret yet for the column as
```
[the remark this answers is not in the document]
Comment (lefessan) [reply cobc/error.c:76]: My intuition is that it's better to display the file as it is, so that it is easy for the developer to visually match the error location in a terminal with the one in an editor.

[the remark this answers is not in the document]
Comment (GitMensch) [reply cobc/error.c:76]: Totally reasonable

[the remark this answers is not in the document]
Comment (lefessan) [reply cobc/error.c:76]: My intuition is that it's better to display the file as it is, so that it is easy for the developer to visually match the error location in a terminal with the one in an editor.

**Review by GitMensch — COMMENTED** [review]
the last commit pushed LGTM

**Review by GitMensch — APPROVED** [review]

**Comment by lefessan** [comment]
I think I added all your suggestions, I will commit it to svn as soon as CI is passed. Thanks !

**Comment by lefessan** [comment]
Merged in SVN

**Comment by GitMensch** [comment]
Sadly that broke C89 support in cobc/error.c: https://ci.appveyor.com/api/buildjobs/vqa49l488as7peee/log

Please adjust to declare the variables at the beginning of the scope.

Apart from that: what do you think of adding a std=c89 to the CI here? That _should_ have likely called that.

**Comment by lefessan** [comment]
Sorry, I didn't see your comment, I am going to fix this problem asap

**Comment by GitMensch** [comment]
Please drop a note if you think you'll have moved around the declarations in the next two hours as I'm doing it otherwise afterwards.

**Comment by GitMensch** [comment]
Fixing C89 and additional changes to the format and some of the details done "upstream" in [svn r5081](https://sourceforge.net/p/gnucobol/code/5081/).

CI PR already referenced.

**Comment by GitMensch** [comment]
Can you have a look for some late-cleanup on this, moving ` -fdiagnostics-plain-output` from `COBC` to `COBOL_FLAGS`, please?
I'd like to keep the command line for NIST, which uses `COBC`, as small as possible.

**Comment by GitMensch** [comment]
In the meantime I've did that adjustment upstream.
